mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java

```diff
@@ -607,12 +635,16 @@ private void trimFlags(Flags flags, MailboxSession session) {
         }, true);
     }
 
-    protected MessageMetaData appendMessageToStore(final MailboxMessage message, MailboxSession session) throws MailboxException {
-        final MessageMapper mapper = mapperFactory.getMessageMapper(session);
+    protected MessageMetaData appendMessageToStore(final MailboxMessage message, final List<Attachment> attachments, MailboxSession session) throws MailboxException {
+        final MessageMapper messageMapper = mapperFactory.getMessageMapper(session);
+        final AttachmentMapper attachmentMapper = mapperFactory.getAttachmentMapper(session);
         return mapperFactory.getMessageMapper(session).execute(new Mapper.Transaction<MessageMetaData>() {
 
             public MessageMetaData run() throws MailboxException {
-                return mapper.add(getMailboxEntity(), message);
+                for (Attachment attachment: attachments) {
+                    attachmentMapper.storeAttachment(attachment);
```

Comment:
we should pass the list to store to let the backend use async apis, don't you think ?